### ISSUE TITLE
A way to disable Elm Optimize Level 2

### ISSUE THREAD

**Issue by tad-lispy** [dillonkearns/elm-pages#288]
As requested by @dillonkearns I'm moving the discussion from <https://discourse.elm-lang.org/t/a-way-to-disable-elm-optimize-level-2-in-elm-pages-build/8281> here, quoting for reference:

Me:

> We run into a bit of a dead end that we described here: <https://github.com/mdgriffith/elm-optimize-level-2/issues/93>. We would love to help resolve it at the root, but it looks like it might take some time. In the meantime we need a workaround to continue working on our project. It seems like using the standard Elm compiler might be the easiest route.
> 
> I quickly glanced at the source code of Elm Pages generator and it seems that elm-optimize-level-2 is a hard dependency. Am I right? If so, would it make sense for it it to be optional? Would you @dillonkearns be open for a PR in this direction? Is there anything I should consider first?
> 
> Probably I'll try to fork it later today and see how far I can go with it. We need to get un-stuck quickly.
> 
> cc: @mdgriffith 
> 
> ps: I'm really grateful for both projects and all the work you and others put into them. Thanks!

Dillon:

> Yeah, I think it would make sense to have an option to disable elm-optimize-level-2. Or another way to put it, to run `elm-pages build` in different modes:
> 
> 1. Debug ( `--debug` )
> 2. Without `--optimize` and without EOL2 (elm-optimize-level-2)
> 3. With `--optimize` but without EOL2
> 4. With `--optimize` and EOL2
> 
> I would love to discuss more, I think a good starting point for the discussion would be to propose what the `elm-pages build` command would look like for each of those 4 modes.
> 
> For example:
> 
> |Number|Description|Command|
> | --- | --- | --- |
> |1|Debug ( `--debug` )|`elm-pages build --debug`|
> |2|Without `--optimize` and without EOL2 (elm-optimize-level-2)|`elm-pages build ???`|
> |3|With `--optimize` but without EOL2|`elm-pages build ???`|
> |4|With `--optimize` and EOL2|`elm-pages build ???`|
> 
> I think a GitHub issue on the `elm-pages` repo would be a better place to discuss those details. Would you mind opening an issue there to continue the conversation?

**Comment by dillonkearns** [comment]
Thanks for the followup!

So one detail to consider here is what the default should be. It's also worth noting that the current default is the fully optimized build (using EOL2). I lean towards the idea of having the default be optimized. It feels safer that way since the most natural thing to do is have a production-ready build that's optimized as much as possible (instead of having to find a magic incantation to make sure you're doing your production builds correctly). And if you need to opt out of it, you can find the appropriate options to do so. That also means that it wouldn't change the existing default, so that seems like an all-around winner to me.

It would also be nice to mirror the familiar options from `elm` as much as possible. I think the `--debug` option is a pretty clear one here. I think it would also be intuitive to keep `--optimize` for option (3) here as well for that reason.

One idea would be to have a CLI flag `--optimize` that can take an argument of a number. I'm not sure if `commander` (the tool that elm-pages uses for CLI flag parsing) supports having a CLI flag that can be used with or without an argument. But if it does, then this seems like a good setup to me:

|Number|Description|Command|
| --- | --- | --- |
|1|Debug ( `--debug` )|`elm-pages build --debug`|
|2|Without `--optimize` and without EOL2 (elm-optimize-level-2)|`elm-pages build --optimize=0`|
|3|With `--optimize` but without EOL2|`elm-pages build --optimize=1`|
|4|With `--optimize` and EOL2|`elm-pages build` (or `--optimize=2`, so 2 would just be the default)|

And then maybe just an error if `--debug` is passed in with anything but `--optimize=0`. What do you think? If that sounds good to you, then I think the only thing remaining would be to see if Commander supports that.

**Comment by tad-lispy** [comment]
Yes, I like it. So basically one will set the optimization level by passing `--optimize=<level>` with default being `2`. And `--optimize` and `--debug` will be mutually exclusive.

Let's see how rusty my Node.js skills are :-)

**Comment by dillonkearns** [comment]
Fixed by #290.
